libraries/standard/http/src/http_client.c

```diff
@@ -247,40 +307,203 @@ static HTTPStatus_t _sendHttpBody( const HTTPTransportInterface_t * pTransport,
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    assert( pTransport != NULL );
+    assert( pTransport->recv != NULL );
+    assert( pBuffer != NULL );
+    assert( pBytesReceived != NULL );
+
+    int32_t transportStatus = pTransport->recv( pTransport->pContext,
+                                                pBuffer,
+                                                bufferLen );
+
+    /* A transport status of less than zero is an error. */
+    if( transportStatus < 0 )
+    {
+        IotLogErrorWithArgs( "Failed to receive HTTP response: Transport recv() "
+                             "returned error: TransportStatus=%d.",
+                             transportStatus );
+        returnStatus = HTTP_NETWORK_ERROR;
+    }
+    else if( (size_t)transportStatus > bufferLen )
+    {
+        /* There is a bug in the transport recv if more bytes are reported
+         * to have been read than the bytes asked for. */
+        IotLogErrorWithArgs( "Failed to receive HTTP response: Transport recv() "
+                             " read more bytes than expected: BytesRead=%d",
+                             transportStatus );
+        returnStatus = HTTP_NETWORK_ERROR;
+    }
+    else if( transportStatus > 0 )
+    {
+        /* Some or all of the specified data was received. */
+        *pBytesReceived = ( size_t ) ( transportStatus );
+        IotLogDebugWithArgs( "Received data from the transport: BytesReceived=%d.",
+                             transportStatus );
+    }
+    else
     {
-        transportStatus = pTransport->send( pTransport->pContext,
-                                            pRequestBodyBuf,
-                                            reqBodyBufLen );
+        /* When a zero is returned from the transport recv it will not be
+         * invoked again. */
+        IotLogDebug( "Transport recv() returned 0. Receiving transport data"
+                     "is complete." );
+    }
+
+    return returnStatus;
+}
 
-        if( transportStatus < 0 )
+/*-----------------------------------------------------------*/
+
+static HTTPStatus_t _getFinalResponseStatus( HTTPParsingState_t parsingState,
+                                             size_t totalReceived,
+                                             size_t responseBufferLen )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    assert( parsingState >= HTTP_PARSING_NONE &&
+            parsingState <= HTTP_PARSING_COMPLETE );
+    assert( totalReceived <= responseBufferLen );
+
+    /* If no parsing occurred, that means network data was never received. */
+    if( parsingState == HTTP_PARSING_NONE )
+    {
+        IotLogErrorWithArgs( "Response not received: Zero returned from "
+                             "transport recv: totalReceived=%d",
+                             totalReceived );
+        returnStatus = HTTP_NO_RESPONSE;
+    }
+    else if( parsingState == HTTP_PARSING_INCOMPLETE )
+    {
+        if( totalReceived == responseBufferLen )
```

Comment:
How about the (corner) case when the server sends incomplete response but the partial response fills the buffer? In that case, the error reported is unsufficient memory though the error could also be partial response.

I suppose that it makes sense to prioritize insufficient memory over partial response. But just stating this case to check whether it was considered

Reply:
That case is moot, we will never be able to deduce it with the current buffer space. A well behaving server is supposed to send the full response and that is an appropriate assumption to make.